utils/hRSMeasurements.m

```diff
@@ -0,0 +1,155 @@
+    cellRSSyms = unique(lsubs(:,1));
+    rssiIndices = repmat((cellRSSyms-1)*K,1,K)+repmat(1:K,length(cellRSSyms),1);
+    rssiIndices = reshape(rssiIndices,numel(rssiIndices),1);
+    meas.AntennaRSRP = NaN(nsf,R);    
+    meas.AntennaRSSI = NaN(nsf,R);    
+    meas.AntennaRSRQ = NaN(nsf,R);  
+    startSubframe = enb.NSubframe;
+    for n = 1:nsf
+        enb.NSubframe = mod(startSubframe+n-1,10);
+        duplexingInfo = lteDuplexingInfo(enb);
+        if (strcmpi(duplexingInfo.SubframeType,'Downlink'))
+            subframe = grid(:,((n-1)*L)+(1:L),:);
+            cellRS = [];
+            for p = 1:length(ports)
+                cellRS = [cellRS; lteCellRS(enb,ports(p))]; %#ok<AGROW>
+            end
+            for r=1:R        
+                subframe(:,:,r) = linearPhaseEqualize(subframe(:,:,r),cellRSIndices,cellRS,K,ksubs,lsubs);
+                meas.AntennaRSRP(n,r) = abs(mean(subframe(cellRSIndices+(r-1)*K*L).*conj(cellRS))*length(ports)).^2;
+                meas.AntennaRSSI(n,r) = sum(abs(subframe(rssiIndices+(r-1)*K*L).^2))/(length(cellRSSyms));   
+                meas.AntennaRSRQ(n,r) = meas.AntennaRSRP(n,r)*double(enb.NDLRB)./meas.AntennaRSSI(n,r);
+            end
+        end
+    end
+    meas.AntennaRSRP = perAntennaMeanFn(meas.AntennaRSRP);
+    meas.AntennaRSSI = perAntennaMeanFn(meas.AntennaRSSI);
+    meas.AntennaRSRQ = perAntennaMeanFn(meas.AntennaRSRQ);
+    idxRSRP = find(meas.AntennaRSRP==max(meas.AntennaRSRP),1);
+    idxRSRQ = find(meas.AntennaRSRQ==max(meas.AntennaRSRQ),1);
+    meas.RSRP = mean(meas.AntennaRSRP(idxRSRP));
+    meas.RSSI = mean(meas.AntennaRSSI(idxRSRQ));    
+    meas.RSRQ = mean(meas.AntennaRSRQ(idxRSRQ));    
+    meas.RSRPdBm = 10*log10(meas.RSRP)+30;
+    meas.RSSIdBm = 10*log10(meas.RSSI)+30; 
+    meas.RSRQdB = 10*log10(meas.RSRQ);
+    
+end
+
+function out = linearPhaseEqualize(in,cellRSIndices,cellRS,K,ksubs,lsubs)
+    
+    nports = size(ksubs,2);
+    cellRSIndices = reshape(cellRSIndices,[],nports);
+    cellRS = reshape(cellRS,[],nports);
+    cellRSSyms = unique(lsubs);
+    nsyms = length(cellRSSyms);
+    thetapoly = zeros(nsyms*nports,2);
+    rms = zeros(nsyms*nports,1);
+    for p = 1:nports
+        lsubsp = lsubs(:,p);
+        for li = 1:nsyms
+            l = cellRSSyms(li);
+            theta = unwrap(angle(in(cellRSIndices(lsubsp==l,p)).*conj(cellRS(lsubsp==l,p))));
+            [thetapoly(li + (p-1)*nsyms,:),s] = polyfit(ksubs(lsubsp==l,p),theta,1);
+            rms(li + (p-1)*nsyms) = s.normr/sqrt(numel(theta)-1);
+        end
+    end
+    threshold = 0.9;
+    if (sum(rms<threshold)>0)
+        thetapoly = mean(thetapoly(rms<threshold,:),1);
+        k = [1:K/2 ((K/2)+2):(K+1)].';
+        theta = polyval(thetapoly,k);
+    else
+        theta = 0;
+    end
+    out = in .* exp(-1i*theta);
+    
+end
+
+function x = perAntennaMeanFn(r)
+
+    x = arrayfun(@(p)mean(r(~isnan(r(:,p)),p)),1:size(r,2));
+    x(isnan(x)) = 0;
+    
+end
```

Comment:
I think this is a build in matlab function, see https://se.mathworks.com/help/lte/examples/reference-signal-measurements-rsrp-rssi-rsrq-for-cell-reselection.html  

Is there a need for a copy here?